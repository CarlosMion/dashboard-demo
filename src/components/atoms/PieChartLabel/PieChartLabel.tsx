import { useMemo } from "react";
import { useTheme } from "@mui/material";

export interface PieChartLabelProps {
  cx: number;
  cy: number;
  midAngle: number;
  innerRadius: number;
  outerRadius: number;
  percent: number;
  name: string;
}

export default function PieChartLabel({
  cx,
  cy,
  midAngle,
  innerRadius,
  outerRadius,
  percent,
  name,
}: PieChartLabelProps) {
  const theme = useTheme();
  const RADIAN = useMemo(() => Math.PI / 180, []);

  const radius = useMemo(
    () => innerRadius + (outerRadius - innerRadius) * 0.6,
    [innerRadius, outerRadius]
  );

  const [x, y] = useMemo(() => {
    const xPos = cx + radius * Math.cos(-midAngle * RADIAN);
    const yPos = cy + radius * Math.sin(-midAngle * RADIAN);
    return [xPos, yPos];
  }, [cx, radius, midAngle, RADIAN, cy]);

  return (
    <text
      x={x}
      y={y}
      fill={theme.palette.common.white}
      textAnchor="middle"
      dominantBaseline="central"
      fontSize="smaller"
    >
      {`${(percent * 100).toFixed(0)}%`}
      <tspan x={x} dy="1.2em">
        {name}
      </tspan>
    </text>
  );
}
