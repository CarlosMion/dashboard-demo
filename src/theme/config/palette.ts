import { alpha, PaletteOptions } from "@mui/material/styles";

declare module "@mui/material/styles/createPalette" {
  interface TypeBackground {
    blackCard: string;
    whiteTransparency: string;
  }

  interface TypeText {
    title: string;
  }

  interface CommonColors {
    depositGreen: string;
    withdrawRed: string;
    orange: string;
    lightGreen: string;
    teal: string;
    mustardYellow: string;
    lightYellow: string;
  }

  interface PaletteOptions {
    commonColors?: Partial<CommonColors>;
  }
}

const palette: PaletteOptions = {
  mode: "light",
  contrastThreshold: 3,
  tonalOffset: 0.2,
  common: {
    black: "#000000",
    white: "#FFFFFF",
  },
  primary: {
    main: "#396AFF",
    light: "#E7EDFF",
    dark: "#1814F3",
  },
  grey: {
    50: alpha("#9199AF", 0.5),
    100: "#B1B1B1",
    200: "#C8D6E2",
    300: "#A3B8C8",
  },
  action: { hover: alpha("#000000", 0.06), selected: alpha("#000000", 0.12) },
  text: {
    primary: "#232323",
    secondary: "#718EBF",
    title: "#343C6A",
  },
  divider: "#DFEAF2",
  background: {
    default: "#F5F7FA",
    paper: "#FFFFFF",
    blackCard: "linear-gradient(0%, #5B5A6F, #000000)",
    whiteTransparency: `linear-gradient(0%, #FFFFFF, ${alpha("#FFFFFF", 0)})`,
  },
  commonColors: {
    depositGreen: "#41D4A8",
    withdrawRed: "#FF5C93",
    orange: "#FC7900",
    lightGreen: "#DCFAF8",
    teal: "#16DBCC",
    mustardYellow: "#FFBB38",
    lightYellow: "#FFF5D9",
  },
};

export default palette;
