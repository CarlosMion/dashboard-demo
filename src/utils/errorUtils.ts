export type ServerError = {
  error: string;
  statusCode?: number;
};

export const isServerErrorWithStatusCode = (
  result: object
): result is ServerError => {
  return "error" in result && "statusCode" in result;
};

export const isServerError = (result: unknown): result is ServerError => {
  return typeof result === "object" && result !== null && "error" in result;
};

type ErrorWithMessage = {
  message: string;
};

export const isErrorWithMessage = (
  error: unknown
): error is ErrorWithMessage => {
  return (
    typeof error === "object" &&
    error !== null &&
    "message" in error &&
    typeof (error as Record<string, unknown>).message === "string"
  );
};

export const toErrorWithMessage = (maybeError: unknown): ErrorWithMessage => {
  if (isErrorWithMessage(maybeError)) return maybeError;

  try {
    return new Error(JSON.stringify(maybeError));
  } catch {
    // fallback in case there's an error stringifying the maybeError
    // like with circular references for example.
    return new Error(String(maybeError));
  }
};

export const getErrorMessage = (error: unknown): string => {
  return toErrorWithMessage(error).message;
};

export class ApiError extends Error {
  statusCode: number;

  constructor(message: string, statusCode: number) {
    super(message);
    this.name = "ApiError";
    this.statusCode = statusCode || 500;
  }
}
