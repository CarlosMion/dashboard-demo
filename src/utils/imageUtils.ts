export const dataURItoBlob = (dataURI: string, fileType: string): Blob => {
  const byteString = atob(dataURI.split(',')[1]);
  const arrayBuffer = new ArrayBuffer(byteString.length);
  const intArray = new Uint8Array(arrayBuffer);

  for (let i = 0; i < byteString.length; i++) {
    intArray[i] = byteString.charCodeAt(i);
  }

  return new Blob([arrayBuffer], { type: fileType });
};

export const dataURItoFile = (
  dataURI: string,
  fileName: string,
  fileType: string
): File => {
  const blob = dataURItoBlob(dataURI, fileType);
  return new File([blob], fileName, { type: blob.type });
};

export const blobToFile = (blob: Blob, fileName: string): File => {
  return new File([blob], fileName, { type: blob.type });
};

export const getImageDimensions = (
  src: string
): Promise<{ width: number; height: number } | null> => {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.src = src;

    img.onload = () =>
      resolve(
        img.width && img.height
          ? { width: img.width, height: img.height }
          : null
      );
    img.onerror = (error) => reject(error);
  });
};

export const isImageDimensionWithinThreshold = (
  actual: number,
  ideal: number,
  threshold: number
): boolean => {
  const isWithinThreshold = actual >= ideal * (1 - threshold);
  return isWithinThreshold;
};
