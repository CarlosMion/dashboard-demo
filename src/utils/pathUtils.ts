export function toCamelCase(str?: string): string | undefined {
  return str?.replace(/-([a-z])/g, (match, letter) => letter.toUpperCase());
}
