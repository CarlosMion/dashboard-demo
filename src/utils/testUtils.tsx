import { render } from "@testing-library/react";
import { ReactNode } from "react";
import AppProviders from "@/providers/AppProviders/AppProviders";

import messages from "@/i18n/messages/en.json";

const customRender = (children: ReactNode) =>
  render(
    <AppProviders
      locale={"en"}
      messages={messages}
      userAgent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
    >
      {children}
    </AppProviders>
  );

export * from "@testing-library/react";

export { customRender as render };
