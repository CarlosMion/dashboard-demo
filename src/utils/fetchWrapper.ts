import { isServerError } from "./errorUtils";

const fetcher = async <T>(
  info: RequestInfo | URL,
  init?: RequestInit
): Promise<T> => {
  let result;
  try {
    const injectedHeaders: Record<string, string> = {};
    if (typeof window === "undefined") {
      const headers = await import("next/headers");
      (await headers.headers()).forEach((value, key) => {
        injectedHeaders[key] = value;
      });
    }

    const buildHeaders: Record<string, string> = {};
    let url = info;
    if (typeof info === "string") {
      url = `${process.env.NEXT_PUBLIC_WEBAPP_API_URL}/${info}`;
    }
    const reqHeaders = {
      "Content-Type": "application/json",
      ...init?.headers,
      ...injectedHeaders,
      ...buildHeaders,
    };
    result = await fetch(url, {
      ...init,
      headers: reqHeaders,
    });
    if (!result.ok) {
      throw new Error("Failed to fetch");
    }
    let parsedResult;
    try {
      parsedResult = await result.json();
    } catch (error) {
      console.log("Failed to parse response, probably not JSON", result, error);
      return result as T;
    }
    if (isServerError(parsedResult)) {
      throw new Error(parsedResult.error);
    }
    return parsedResult as T;
  } catch (error) {
    throw error;
  }
};

export default fetcher;
