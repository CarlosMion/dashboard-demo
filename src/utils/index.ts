import { isEdge } from "./userAgentUtils";
import { maskCardNumber } from "./stringUtils";
import { toCamelCase } from "./pathUtils";
import fetcher from "./fetchWrapper";
import {
  isServerErrorWithStatusCode,
  ApiError,
  getErrorMessage,
  isErrorWithMessage,
  isServerError,
  toErrorWithMessage,
} from "./errorUtils";
import {
  blobToFile,
  dataURItoBlob,
  dataURItoFile,
  getImageDimensions,
  isImageDimensionWithinThreshold,
} from "./imageUtils";
import {
  createImage,
  getRadianAngle,
  rotateSize,
  getCroppedImg,
} from "./cropImageUtils";
export {
  fetcher,
  isEdge,
  toCamelCase,
  isServerErrorWithStatusCode,
  ApiError,
  getErrorMessage,
  isErrorWithMessage,
  isServerError,
  toErrorWithMessage,
  blobToFile,
  dataURItoBlob,
  dataURItoFile,
  getImageDimensions,
  isImageDimensionWithinThreshold,
  getCroppedImg,
  rotateSize,
  getRadianAngle,
  createImage,
  maskCardNumber,
};
