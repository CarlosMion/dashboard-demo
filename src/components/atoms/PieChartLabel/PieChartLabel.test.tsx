import { render, screen } from "@/utils/testUtils";
import PieChartLabel from "./PieChartLabel";
import palette from "@/theme/config/palette";

const mockHtmlElements = () => {
  const originalCreateElement = document.createElement.bind(document);
  jest.spyOn(document, "createElement").mockImplementation((tagName) => {
    if (tagName === "text" || tagName === "tspan") {
      return originalCreateElement("span");
    }
    return originalCreateElement(tagName);
  });
};

describe("PieChartLabel", () => {
  beforeAll(() => {
    mockHtmlElements();
  });
  it("renders correctly and matches snapshot", () => {
    const { container } = render(
      <PieChartLabel
        cx={0}
        cy={0}
        midAngle={0}
        innerRadius={0}
        outerRadius={0}
        percent={0}
        name={""}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("displays the correct label text", () => {
    const name = "Test Label";
    render(
      <PieChartLabel
        cx={0}
        cy={0}
        midAngle={0}
        innerRadius={0}
        outerRadius={0}
        percent={0}
        name={name}
      />
    );
    expect(screen.getByText(name)).toBeInTheDocument();
  });

  it("displays the correct value", () => {
    render(
      <PieChartLabel
        cx={0}
        cy={0}
        midAngle={0}
        innerRadius={0}
        outerRadius={0}
        percent={25}
        name={""}
      />
    );
    expect(screen.getByText("25%")).toBeInTheDocument();
  });

  it("applies the correct color to the indicator", () => {
    const { container } = render(
      <PieChartLabel
        cx={0}
        cy={0}
        midAngle={0}
        innerRadius={0}
        outerRadius={0}
        percent={0}
        name={""}
      />
    );
    const indicator = container.firstChild;
    expect(indicator).toHaveAttribute("fill", palette.common?.white);
  });

  it("handles zero value correctly", () => {
    render(
      <PieChartLabel
        cx={0}
        cy={0}
        midAngle={0}
        innerRadius={0}
        outerRadius={0}
        percent={0}
        name={""}
      />
    );
    expect(screen.getByText("0%")).toBeInTheDocument();
  });
});
