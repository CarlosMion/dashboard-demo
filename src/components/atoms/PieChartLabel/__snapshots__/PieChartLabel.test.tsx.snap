// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChartLabel renders correctly and matches snapshot 1`] = `
<span
  dominant-baseline="central"
  fill="#FFFFFF"
  font-size="smaller"
  text-anchor="middle"
  x="0"
  y="0"
>
  0%
  <span
    dy="1.2em"
    x="0"
  />
</span>
`;
